### DISCUSSION TITLE
Sprinklers problem

### DISCUSSION THREAD

**Discussion by pielle63** [HomeSpan/HomeSpan#473] [Q&A]
Hello,
HomeSpan is a really great library, I've extensively used it to control a lot of things and they work like charme
Now I'm working on the irrigation system: the Valve accessory works as expected but... there is no way to get the IrrigationSystem accessory working. seems to be a useless icon in House App, almost insensitive to what is coded.
I went through the HomeSpan issues and discussions looking for some information - I've found some useful results but nothing definitive for scheduling Valve operations. I also downloaded a copule of examples from HomeSpan but both behaves like mine. 
Has anyone managed to figure it out?

**Comment by ghost** [comment]
I do not really understand what your problem is.
I tried this example:
https://github.com/HomeSpan/HomeSpan/blob/3ce027cf466b39d9834d55c0bfad902d58a703e5/examples/17-LinkedServices/17-LinkedServices.ino
and it was working fine.
Then I switched in line 120 from Valve to IrrigationSystem and added Serial.println in the loop.
I see the device and when I change the value I can see the output in my console.

What exactly do you expect to happen?
Without knowing too much what an IrrigationSystem does, I can see the icon, I can open it and I can switch it from on to off, just as I expected.

**Comment by HomeSpan** [comment]
Sorry, I somehow missed this question from a few days ago.  In any case, below please find a *Reference Irrigation System Sketch* I created quite a while back --- I just retested under iOS 16.2 and it still seems to work fine.

The sketch shows how to implement an Irrigation System with 4 Valves that can be separately controlled, renamed, and enabled/disabled.  The duration of each valve can be set in the Home App from the settings screen for the device.  The Home App has a fixed set of choices (5 minutes to 1 hour), though your device can specify any duration it would like (the sketch does not currently have a hook to do this).

Comments in the code indicate where I needed to deviate from HAP-R2 specs in order to the get device to work properly with the Home App.  Would be nice if Apple provided updated HAP specs to the public.

```C++
/*********************************************************************************
 *  MIT License
 *  
 *  Copyright (c) 2023 Gregg E. Berman
 *  
 *  https://github.com/HomeSpan/HomeSpan
 *  
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *  
 *  The above copyright notice and this permission notice shall be included in all
 *  copies or substantial portions of the Software.
 *  
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 *  
 ********************************************************************************/

 
///////////////////////////////////////////////////////
//                                                   //
//   HomeSpan Reference Sketch: Irrigation Service   //
//                                                   //
///////////////////////////////////////////////////////
 
#include "HomeSpan.h" 

struct Head : Service::Valve {

  SpanCharacteristic *active=new Characteristic::Active(0);
  SpanCharacteristic *inUse=new Characteristic::InUse(0);
  SpanCharacteristic *enabled = new Characteristic::IsConfigured(1,true);
  SpanCharacteristic *setDuration = new Characteristic::SetDuration(300);
  SpanCharacteristic *remainingDuration = new Characteristic::RemainingDuration(0);
  SpanCharacteristic *name;

  Head(const char *headName) : Service::Valve() {
    new Characteristic::ValveType(1);
    name=new Characteristic::ConfiguredName(headName,true);     // This Characteristic was introduced for TV Services, but works well here
    enabled->addPerms(PW);                                      // Adding "PW" to the IsConfigured Characteristic allws for enabling/disabling valves
  }

  boolean update() override {
    
    if(enabled->updated()){
      if(enabled->getNewVal()){
        Serial.printf("%s value ENABLED\n",name->getString());
      } else {
        Serial.printf("%s value DISABLED\n",name->getString());          
        if(active->getVal()){
          active->setVal(0);
          remainingDuration->setVal(0);
          Serial.printf("%s is CLOSING\n",name->getString());          
        }
      }
    }

    if(active->updated()){
      if(active->getNewVal()){
        Serial.printf("%s valve is OPENING\n",name->getString());
        remainingDuration->setVal(setDuration->getVal());
      } else {
        Serial.printf("%s valve is CLOSING\n",name->getString());
        remainingDuration->setVal(0);
      }
    }
    
    return(true);
  }

  void loop() override {
    if(active->getVal()){
      int remainingTime=setDuration->getVal()-active->timeVal()/1000;
         
      if(remainingTime<=0){
        Serial.printf("%s valve is CLOSING (%d-second timer is complete)\n",name->getString(),setDuration->getVal());
        active->setVal(0);
        remainingDuration->setVal(0);
      } else

      if(remainingTime<remainingDuration->getVal()){
        remainingDuration->setVal(remainingTime,false);
      }
    }

    // Below we simulate valves that take 5 seconds to open/close so that In Use follows Active by 5 seconds
    // The Home App will accurately reflect this intermediate state and show "Waiting..." when a value of initially
    // activated
    
    if(active->timeVal()>5000 && active->getVal()!=inUse->getVal()){
      inUse->setVal(active->getVal());
      Serial.printf("%s value is %s\n",name->getString(),inUse->getVal()?"OPEN":"CLOSED");
    }
  }

};

////////////////////////////////////////////////////////////////////////

struct Sprinkler : Service::IrrigationSystem {

  SpanCharacteristic *active=new Characteristic::Active(0);             // HomeKit requires this Characteristic, but it has no effect in Home App
  SpanCharacteristic *programMode=new Characteristic::ProgramMode(0);   // HomeKit requires this Characteristic, but it is mostly for information purposeses only in the Home App

  // NOTE: According to HAP-R2, the In Use Characteristic is also required for the Irrigation System Service.
  // However, adding this Characteristic seems to break the Home App.  Seems that the HAP-R2 spec is out of date
  // and that the Home App determines whether the Irrigation System is In Use simply by noting that one or more
  // linked Valves are In Use.
  //
  // Recommendation: Do NOT instatiate Characteristic::InUse for the Irrigation Service
};

////////////////////////////////////////////////////////////////////////

void setup() {

  Serial.begin(115200);
  
  homeSpan.begin(Category::Sprinklers,"HomeSpan Sprinklers");

  new SpanAccessory();                                  
    new Service::AccessoryInformation();  
      new Characteristic::Identify();                           
                   
    (new Sprinkler())
      ->addLink(new Head("Head 1"))
      ->addLink(new Head("Head 2"))
      ->addLink(new Head("Head 3"))
      ->addLink(new Head("Head 4"))
      ;

} // end of setup()

////////////////////////////////////////////////////////////////////////

void loop(){
  
  homeSpan.poll();
  
} // end of loop()

//////////////////////////////////////
```

**Comment by pielle63** [comment]
thank you for your answer, probably I badly expressed myself.  I would like to implement an automatation/programming of the sprinkler e.g. automatically start waterting valve 1 at time xx:xx and stop it after nn minutes;  start waterting valve 2 at time yy:yy and stop it after mm minutes and so on.
This currently is not allowed and I was tryng to understand if this functionality could be provided by the IrrigationSystem service.

**Comment by HomeSpan** [comment]
Yes, you can definitely do this, but the implementation needs to be within the code.  Apple does not allow water-operated devices to be scheduled via the Home App in recognition of the fact that sometimes HomeKit automation/scheduling doesn't always work.  If you scheduled a light to turn off at dawn and it does not, probably no big deal.  But if you schedule you water to turn off in 5 minutes and HomeKit fails to send a shut-off to the device, you can be left with water running for hours.

That means that all scheduling needs to be done outside of the HomeKit framework by the device, with status updates as to which valves are open/closed sent to HomeKit so you can monitor progress in the Home App, as well as manually override.

HomeSpan does not have any libraries to perform generic scheduling but you should be able to create your own code and have the sketch run it from the main `loop()`, provided you use `setVal()` as needed to let HomeKit know when you turn on/off valves.

**Reply by HomeSpan** [reply]
@pielle63, I just created a new standalone repository of [HomeSpan Reference Sketches](https://github.com/HomeSpan/HomeSpanReferenceSketches) showcasing some of the more complex HomeKit Services.  One of the sketches is for an Irrigation System that includes a scheduled program.  You may find it helpful in designing your own scheduling system.

**Reply by pielle63** [reply]
thank you so much HomeSpan: it is exaclty what I was looking for!
